File: src/commands/pb.ts
import { SlashCommandBuilder, ChatInputCommandInteraction, EmbedBuilder } from "discord.js";
import { reply } from "../lib/functions/discord";
import { parseTime } from "../lib/functions/util";
import { validateTarget } from "../lib/functions/schnose";
import { getPlace } from "gokz.js";
import userSchema from "../lib/schemas/user";
import modeMap from "gokz.js/lib/api";
import SchnoseBot from "src/classes/Schnose";
import { pb_wasm } from "../../rust/pkg/gokz_wasm";
import * as W from "src/lib/types/wasm";

export default {
	data: new SlashCommandBuilder()
		.setName("pb")
		.setDescription("Check a player's personal best on a map.")
		.addStringOption((o) => o.setName("map").setDescription("Specify a map.").setRequired(true))
		.addStringOption((o) =>
			o.setName("mode").setDescription("Specify a mode.").setChoices(
				{
					name: "KZT",
					value: "kz_timer"
				},
				{
					name: "SKZ",
					value: "kz_simple"
				},
				{
					name: "VNL",
					value: "kz_vanilla"
				}
			)
		)
		.addStringOption((o) => o.setName("target").setDescription("Specify a target.")),

	async execute(interaction: ChatInputCommandInteraction, client: SchnoseBot) {
		await interaction.deferReply();

		const inputMap = interaction.options.getString("map")!;
		const inputMode = interaction.options.getString("mode") || null;
		const inputTarget = interaction.options.getString("target") || null;

		let mode: string;
		if (inputMode) mode = inputMode;
		else {
			const userDB = await userSchema.find({ discordID: interaction.user.id });
			if (!userDB[0]?.mode)
				return reply(interaction, {
					content: "You must either specify a mode or set a default value using `/mode`."
				});
			else mode = userDB[0].mode;
		}

		const targetValidation = await validateTarget(interaction, inputTarget);
		if (!targetValidation.success) return reply(interaction, { content: targetValidation.error });

		const request = await pb_wasm(inputMap, mode, targetValidation.data!.value!);

		const result: W.pb_wasm[] = [null, null];
		try {
			const temp = JSON.parse(request);
			for (let i = 0; i < temp.length; i++) {
				try {
					result[i] = JSON.parse(temp[i]) as W.pb_wasm;
				} catch (_) {}
			}
		} catch (_) {
			return reply(interaction, { content: request });
		}

		let tpPlace: any = null; // eslint-disable-line
		let proPlace: any = null; // eslint-disable-line
		if (result[0]?.map_name) tpPlace = await getPlace(result[0]);
		if (result[1]?.map_name) proPlace = await getPlace(result[1]);

		const embed = new EmbedBuilder()
			.setColor([116, 128, 194])
			.setTitle(
				`[PB] ${
					result[0]?.player_name || result[1]?.player_name
						? `${result[0]?.player_name || result[1]?.player_name} on ${
								result[0]?.map_name || result[1]?.map_name // eslint-disable-line
						  }` // eslint-disable-line
						: `${result[0]?.map_name || result[1]?.map_name}`
				}`
			)
			.setURL(`https://kzgo.eu/maps/${result[0]?.map_name || result[1]?.map_name}`)
			.setThumbnail(
				`https://raw.githubusercontent.com/KZGlobalTeam/map-images/master/images/${
					result[0]?.map_name || result[1]?.map_name
				}.jpg`
			)
			.addFields([
				{
					name: "TP",
					value: `${result[0]?.time ? parseTime(result[0].time) : "😔"} ${
						tpPlace && tpPlace.success ? `(#${tpPlace?.data})` : `${result[0] ? "?" : ""}`
					}`,
					inline: true
				},
				{
					name: "PRO",
					value: `${result[1]?.time ? parseTime(result[1].time) : "😔"} ${
						proPlace && proPlace.success ? `(#${proPlace?.data})` : `${result[1] ? "?" : ""}`
					}`,
					inline: true
				}
			])
			.setFooter({
				text: `(͡ ͡° ͜ つ ͡͡°)7 | Mode: ${modeMap.get(mode)}`,
				iconURL: client.icon
			});

		return reply(interaction, {
			embeds: [embed]
		});
	}
};

File: src/commands/pb.rs
use {
	crate::{
		events::slash_commands::{
			GlobalState,
			InteractionResponseData::{self, *},
		},
		schnose::Target,
		util::*,
		db::retrieve_mode,
		gokz::{self, *},
	},
	futures::future::join_all,
	gokz_rs::{prelude::*, global_api::*},
	itertools::Itertools,
	serenity::{
		builder::{CreateApplicationCommand, CreateEmbed},
		model::prelude::command::CommandOptionType,
	},
};

pub(crate) fn register(cmd: &mut CreateApplicationCommand) -> &mut CreateApplicationCommand {
	return cmd
		.name("pb")
		.description("Check a player's personal best on a map.")
		.create_option(|opt| {
			opt.kind(CommandOptionType::String)
				.name("map_name")
				.description("Specify a map.")
				.required(true)
		})
		.create_option(|opt| {
			opt.kind(CommandOptionType::String)
				.name("mode")
				.description("Choose a mode.")
				.add_string_choice("KZT", "kz_timer")
				.add_string_choice("SKZ", "kz_simple")
				.add_string_choice("VNL", "kz_vanilla")
				.required(false)
		})
		.create_option(|opt| {
			opt.kind(CommandOptionType::String)
				.name("player")
				.description("Specify a player.")
				.required(false)
		});
}

pub(crate) async fn execute(
	state: &mut GlobalState<'_>,
) -> anyhow::Result<InteractionResponseData> {
	state.defer().await?;

	let map_name = state.get::<String>("map_name").expect("This option is marked as `required`.");
	let target = Target::from(state.get::<String>("player"));
	let player = match target.to_player(state.user, state.db).await {
		Ok(player) => player,
		Err(why) => {
			log::warn!("[{}]: {} => {:?}", file!(), line!(), why);
			return Ok(Message(why.to_string()));
		},
	};
	let mode = match state.get::<String>("mode") {
		Some(mode_name) => Mode::from_str(&mode_name).expect("This must be valid at this point."),
		None => match retrieve_mode(state.user, state.db).await {
			Ok(mode) => mode,
			Err(why) => {
				log::warn!("[{}]: {} => {:?}", file!(), line!(), why);
				return Ok(Message(why.to_string()));
			},
		},
	};

	let global_maps = match get_maps(&state.req_client).await {
		Ok(maps) => maps,
		Err(why) => {
			log::warn!("[{}]: {} => {:?}", file!(), line!(), why);
			return Ok(Message(why.tldr));
		},
	};

	let map = match is_global(&MapIdentifier::Name(map_name), &global_maps).await {
		Ok(map) => map,
		Err(why) => {
			log::warn!("[{}]: {} => {:?}", file!(), line!(), why);
			return Ok(Message(why.tldr));
		},
	};

	let map_identifier = MapIdentifier::Name(map.name.clone());

	let (tp, pro) = join_all([
		get_pb(&player, &map_identifier, &mode, true, 0, &state.req_client),
		get_pb(&player, &map_identifier, &mode, false, 0, &state.req_client),
	])
	.await
	.into_iter()
	.collect_tuple()
	.expect("This cannot fail, look 6 lines up.");

	if let (&Err(_), &Err(_)) = (&tp, &pro) {
		return Ok(Message("No PBs found 😔.".into()));
	}

	let player_name = get_player_name((&tp, &pro));
	let (place_tp, place_pro) = (
		gokz::get_place(&tp, &state.req_client).await,
		gokz::get_place(&pro, &state.req_client).await,
	);
	let links = (gokz::get_replay_link(&tp).await, gokz::get_replay_link(&pro).await);

	let mut embed = CreateEmbed::default()
		.colour(state.colour)
		.title(format!("[PB] {} on {} (T{})", &player_name, &map.name, &map.difficulty))
		.url(format!(
			"https://kzgo.eu/maps/{}?{}=",
			&map.name,
			&mode.to_fancy().to_lowercase()
		))
		.thumbnail(state.thumbnail(&map.name))
		.field(
			"TP",
			format!(
				"{} {}",
				match &tp {
					Ok(rec) => format_time(rec.time),
					_ => String::from("😔"),
				},
				place_tp
			),
			true,
		)
		.field(
			"PRO",
			format!(
				"{} {}",
				match &pro {
					Ok(rec) => format_time(rec.time),
					_ => String::from("😔"),
				},
				place_pro
			),
			true,
		)
		.footer(|f| f.text(format!("Mode: {}", mode.to_fancy())).icon_url(&state.icon))
		.to_owned();

	attach_replay_links(&mut embed, links);

	return Ok(Embed(embed));
}
